core/vote/vote_pool.go

```diff
@@ -0,0 +1,401 @@
+	scope     event.SubscriptionScope
+
+	receivedVotes mapset.Set
+
+	curVotes    map[common.Hash]*VoteBox
+	futureVotes map[common.Hash]*VoteBox
+
+	curVotesPq    *votesPriorityQueue
+	futureVotesPq *votesPriorityQueue
+
+	chainHeadCh  chan core.ChainHeadEvent
+	chainHeadSub event.Subscription
+
+	votesCh chan *types.VoteEnvelope
+
+	engine consensus.PoSA
+}
+
+type votesPriorityQueue []*types.VoteData
+
+func NewVotePool(chainconfig *params.ChainConfig, chain *core.BlockChain, engine consensus.PoSA) *VotePool {
+	votePool := &VotePool{
+		chain:         chain,
+		chainconfig:   chainconfig,
+		receivedVotes: mapset.NewSet(),
+		curVotes:      make(map[common.Hash]*VoteBox),
+		futureVotes:   make(map[common.Hash]*VoteBox),
+		curVotesPq:    &votesPriorityQueue{},
+		futureVotesPq: &votesPriorityQueue{},
+		chainHeadCh:   make(chan core.ChainHeadEvent, chainHeadChanSize),
+		votesCh:       make(chan *types.VoteEnvelope, voteBufferForPut),
+		engine:        engine,
+	}
+
+	// Subscribe events from blockchain and start the main event loop.
+	votePool.chainHeadSub = votePool.chain.SubscribeChainHeadEvent(votePool.chainHeadCh)
+	if posa, ok := engine.(consensus.PoSA); ok {
+		posa.SetVotePool(votePool)
+	}
+
+	go votePool.loop()
+	return votePool
+}
+
+// loop is the vote pool's main even loop, waiting for and reacting to outside blockchain events and votes channel event.
+func (pool *VotePool) loop() {
+	for {
+		select {
+		// Handle ChainHeadEvent.
+		case ev := <-pool.chainHeadCh:
+			if ev.Block != nil {
+				latestBlockNumber := ev.Block.NumberU64()
+				pool.prune(latestBlockNumber)
+				pool.transferVotesFromFutureToCur(ev.Block.Header())
+			}
+		case <-pool.chainHeadSub.Err():
+			return
+
+		// Handle votes channel and put the vote into vote pool.
+		case vote := <-pool.votesCh:
+			pool.putIntoVotePool(vote)
+		}
+	}
+}
+
+func (pool *VotePool) PutVote(vote *types.VoteEnvelope) {
+	pool.votesCh <- vote
+}
+
+func (pool *VotePool) putIntoVotePool(vote *types.VoteEnvelope) bool {
+	targetNumber := vote.Data.TargetNumber
+	targetHash := vote.Data.TargetHash
+	header := pool.chain.CurrentBlock().Header()
+	headNumber := header.Number.Uint64()
+
+	voteData := &types.VoteData{
+		TargetNumber: targetNumber,
+		TargetHash:   targetHash,
+	}
+
+	var votes map[common.Hash]*VoteBox
+	var votesPq *votesPriorityQueue
+	isFutureVote := false
+
+	voteBlock := pool.chain.GetHeaderByHash(targetHash)
+	if voteBlock == nil {
+		votes = pool.futureVotes
+		votesPq = pool.futureVotesPq
+		isFutureVote = true
+	} else {
+		votes = pool.curVotes
+		votesPq = pool.curVotesPq
+	}
+
+	voteHash := vote.Hash()
+	if ok := pool.basicVerify(vote, headNumber, votes, isFutureVote, voteHash); !ok {
+		return false
+	}
+
+	if !isFutureVote {
+		// Verify if the vote comes from valid validators based on voteAddress (BLSPublicKey), only verify curVotes here, will verify futureVotes in transfer process.
+		if posa, ok := pool.engine.(consensus.PoSA); ok {
```

Comment:
It is already a `consensus.PoSA`, no need to do assert.

Reply:
Done